File: modules/data/src/main/java/org/gjt/jclasslib/bytecode/WideBranchInstruction.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.bytecode;

import org.gjt.jclasslib.io.ByteCodeInput;
import org.gjt.jclasslib.io.ByteCodeOutput;

import java.io.IOException;

/**
    Describes an instruction that is followed by an immediate int.
 
    @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
*/
public class WideBranchInstruction extends AbstractBranchInstruction {

    /**
        Constructor.
        @param opcode the opcode.
     */
    public WideBranchInstruction(Opcode opcode) {
        super(opcode); 
    }

    /**
        Constructor.
        @param opcode the opcode.
        @param branchOffset the immediate int value.
     */
    public WideBranchInstruction(Opcode opcode, int branchOffset) {
        super(opcode, branchOffset);
    }
    
    public int getSize() {
        return super.getSize() + 4;
    }

    public void read(ByteCodeInput in) throws IOException {
        super.read(in);

        setBranchOffset(in.readInt());
    }

    public void write(ByteCodeOutput out) throws IOException {
        super.write(out);

        out.writeInt(getBranchOffset());
    }
    
}

File: modules/data/src/main/kotlin/org/gjt/jclasslib/bytecode/WideBranchInstruction.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.bytecode

import org.gjt.jclasslib.io.ByteCodeInput
import org.gjt.jclasslib.io.ByteCodeOutput

/**
 * Describes an instruction that is followed by an immediate int.
 */
class WideBranchInstruction @JvmOverloads constructor(opcode: Opcode, branchOffset: Int = 0) : AbstractBranchInstruction(opcode, branchOffset) {

    override val size: Int
        get() = super.size + 4

    override fun read(input: ByteCodeInput) {
        super.read(input)

        branchOffset = input.readInt()
    }

    override fun write(output: ByteCodeOutput) {
        super.write(output)

        output.writeInt(branchOffset)
    }

}
